components/raftstore/src/store/fsm/store.rs

```diff
@@ -120,6 +120,8 @@ pub struct StoreMeta {
     pub region_ranges: BTreeMap<Vec<u8>, u64>,
     /// region_id -> region
     pub regions: HashMap<u64, Region>,
+    /// region_id -> LocalLeaderInfo
+    pub region_leaders: HashMap<u64, LocalLeaderInfo>,
```

Comment:
what's LocalLeaderInfo for? Is the bool not enough?

Reply:
> what's LocalLeaderInfo for? Is the bool not enough?

A bool is enough for the issue we met, but it would be better if we can put the leader info is in NotLeader error. 
```rust
#[error("peer is not leader for region {0}, leader may {1:?}")]
    NotLeader(u64, Option<metapb::Peer>),
```

> Also this StoreMeta is shared by many code

I've struggled with this before, the StoreMeta is shared by many code currently. 

However, after reading the related code, I think some fields are suitable for storing in StoreMeta and some are not. Actually, I think the following fields are suitable, and so does the `region_leaders` info.
https://github.com/tikv/tikv/blob/6b3ca45c8e828d5ef8b36492791237ea9d62cfca/components/raftstore/src/store/fsm/store.rs#L118-L123
And maybe the following are not.
https://github.com/tikv/tikv/blob/6b3ca45c8e828d5ef8b36492791237ea9d62cfca/components/raftstore/src/store/fsm/store.rs#L151

-------------------

I think a better way is to remove these unsuitable fields instead of move suitable fields like `region_leaders` into another struct. Otherwise, we will have another struct just like StoreMeta, and each time StoreMeta is updated, we need to update the new struct.

Reply:
After some investigation, I do think that we just need a hashset to store these region leader ids. And the implementation can be simplified. StoreMeta is not needed :)

Reply:
After some investigation, I do think that we just need a hashset to store these region leader ids. And the implementation can be simplified. StoreMeta is not needed :)

Reply:
done